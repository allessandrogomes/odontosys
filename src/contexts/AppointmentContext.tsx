import { createContext, useReducer, useContext, ReactNode } from "react"

type Appointment = {
    patientName: string | null
    patientId: number | null
    dentistName: string | null
    dentistId: number | null
    scheduledAt: string | null
    endsAt: string | null
    durationMinutes: number | null
    procedure: string | null
    step: number
}

type State = Appointment

type Action =
    | { type: "SET_PATIENT"; payload: { id: number | null; name: string | null } }
    | { type: "SET_DENTIST"; payload: { id: number | null; name: string | null } }
    | { type: "SET_SCHEDULE"; payload: { scheduledAt: string | null; endsAt: string | null } }
    | { type: "SET_PROCEDURE"; payload: { procedure: string | null; durationMinutes: number | null }  }
    | { type: "SET_STEP"; payload: number }
    | { type: "RESET_APPOINTMENT" }

const initialState: State = {
    patientName: null,
    patientId: null,
    dentistName: null,
    dentistId: null,
    scheduledAt: null,
    endsAt: null,
    durationMinutes: null,
    procedure: null,
    step: 1
}

function reducer(state: State, action: Action): State {
    switch (action.type) {
        case "SET_PATIENT":
            return {
                ...state,
                patientId: action.payload.id,
                patientName: action.payload.name,
            }
        case "SET_DENTIST":
            return {
                ...state,
                dentistId: action.payload.id,
                dentistName: action.payload.name,
            }
        case "SET_SCHEDULE":
            return {
                ...state,
                scheduledAt: action.payload.scheduledAt,
                endsAt: action.payload.endsAt,
            }
        case "SET_PROCEDURE":
            return {
                ...state,
                procedure: action.payload.procedure,
                durationMinutes: action.payload.durationMinutes,
            }
        case "SET_STEP":
            return {
                ...state,
                step: action.payload
            }
        case "RESET_APPOINTMENT":
            return initialState
        default:
            return state
    }
}

const AppointmentContext = createContext<{
    state: State
    dispatch: React.Dispatch<Action>
} | undefined>(undefined)

export function AppointmentProvider({ children }: { children: ReactNode }) {
    const [state, dispatch] = useReducer(reducer, initialState)

    return (
        <AppointmentContext.Provider value={{ state, dispatch }}>
            {children}
        </AppointmentContext.Provider>
    )
}

export function useAppointmentContext() {
    const context = useContext(AppointmentContext)
    if (!context) {
        throw new Error("useAppointment must be used within an AppointmentProvider")
    }
    return context
}