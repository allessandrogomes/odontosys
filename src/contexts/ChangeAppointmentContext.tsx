import { createContext, useReducer, useContext, ReactNode } from "react"

type ChangeAppointmentState = {
    appointments: IAppointment[] | [] | null
    selectedAppointment: IAppointment | null
    step: number
}

type State = ChangeAppointmentState

type Action =
    | { type: "SET_APPOINTMENTS"; payload: IAppointment[] | [] | null }
    | { type: "SET_SELECTED_APPOINTMENT"; payload: IAppointment | null }
    | { type: "SET_STEP"; payload: number }

const initialState: State = {
    appointments: null,
    selectedAppointment: null,
    step: 1
}

function reducer(state: State, action: Action): State {
    switch (action.type) {
        case "SET_APPOINTMENTS":
            return {
                ...state,
                appointments: action.payload
            }
        case "SET_SELECTED_APPOINTMENT":
            return {
                ...state,
                selectedAppointment: action.payload
            }
        case "SET_STEP":
            return {
                ...state,
                step: action.payload
            }
        default:
            return state
    }
}

const ChangeAppointmentContext = createContext<{
    state: State
    dispatch: React.Dispatch<Action>
} | undefined>(undefined)

export function ChangeAppointmentProvider({ children }: { children: ReactNode }) {
    const [state, dispatch] = useReducer(reducer, initialState)

    return (
        <ChangeAppointmentContext.Provider value={{ state, dispatch }}>
            {children}
        </ChangeAppointmentContext.Provider>
    )
}

export function useChangeAppointmentContext() {
    const context = useContext(ChangeAppointmentContext)
    if (!context) {
        throw new Error("useChangeAppointmentContext must be used within and ChangeAppointmentProvider")
    }
    return context
}