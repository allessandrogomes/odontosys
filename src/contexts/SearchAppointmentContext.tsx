import { createContext, useReducer, useContext, ReactNode } from "react"

type SearchAppointmentState = {
    appointments: IAppointment[] | [] | null
    selectedAppointment: IAppointment | null
    step: number
}

type State = SearchAppointmentState

type Action =
    | { type: "SET_APPOINTMENTS"; payload: IAppointment[] | [] | null }
    | { type: "SET_SELECTED_APPOINTMENT"; payload: IAppointment | null }
    | { type: "SET_STEP"; payload: number }

const initialState: State = {
    appointments: null,
    selectedAppointment: null,
    step: 1
}

function reducer(state: State, action: Action): State {
    switch (action.type) {
        case "SET_APPOINTMENTS":
            return {
                ...state,
                appointments: action.payload
            }
        case "SET_SELECTED_APPOINTMENT":
            return {
                ...state,
                selectedAppointment: action.payload
            }
        case "SET_STEP":
            return {
                ...state,
                step: action.payload
            }
        default:
            return state
    }
}

const SearchAppointmentContext = createContext<{
    state: State
    dispatch: React.Dispatch<Action>
} | undefined>(undefined)

export function SearchAppointmentProvider({ children }: { children: ReactNode }) {
    const [state, dispatch] = useReducer(reducer, initialState)

    return (
        <SearchAppointmentContext.Provider value={{ state, dispatch }}>
            {children}
        </SearchAppointmentContext.Provider>
    )
}

export function useSearchAppointmentContext() {
    const context = useContext(SearchAppointmentContext)
    if (!context) {
        throw new Error("useSearchAppointment must be used within an SearchAppointmentProvider")
    }
    return context
}