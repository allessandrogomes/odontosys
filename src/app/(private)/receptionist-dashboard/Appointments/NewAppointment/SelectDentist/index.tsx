import styles from "./styles.module.scss"
import Label from "@/components/ui/Label"
import Button from "@/components/ui/Button"
import { ArrowLeft, ArrowRight, OctagonX } from "lucide-react"
import Spinner from "@/components/ui/Spinner"
import useSWR from "swr"
import FeedbackMessage from "@/components/ui/FeedbackMessage"
import { useAppointmentContext } from "@/contexts/AppointmentContext"
import fetcher from "@/services/dentist/getDentistByProcedure"

export default function SelectDentist() {
    const { state, dispatch } = useAppointmentContext()
    const { data: dentists, error, isLoading } = useSWR<IDentist[]>(`/api/dentist?procedure=${encodeURIComponent(state.procedure!)}`, fetcher, { revalidateOnFocus: false })

    function handleSelectDentist(dentist: IDentist) {
        dispatch({
            type: "SET_DENTIST",
            payload: { id: dentist.id, name: dentist.name }
        })
    }

    function handleBack() {
        dispatch({ type: "SET_STEP", payload: state.step - 1 })
        dispatch({ type: "SET_DENTIST", payload: { id: null, name: null } }) // Reseta dentista selecionado
    }

    function handleNext() {
        dispatch({ type: "SET_STEP", payload: state.step + 1 })
    }

    return (
        <div className={styles.box}>
            <div className={styles.selectView}>
                <Label text="Escolha o Dentista" />
                {isLoading ? <Spinner />
                    : error ? <FeedbackMessage message={error} icon={<OctagonX />} />
                        : (
                            <div className={styles.containerDentists}>
                                {dentists && dentists.map((dentist: IDentist) => (
                                    <div
                                        role="button"
                                        tabIndex={0}
                                        aria-label={`Selecionar dentista ${dentist.name}`}
                                        className={`${styles.dentist} ${state.dentistId === dentist.id ? styles.selected : ""}`}
                                        onClick={() => handleSelectDentist(dentist)}
                                        key={dentist.id}
                                    >
                                        <p>Dr. {dentist.name}</p>
                                    </div>
                                ))}
                            </div>
                        )}
                <div className={styles.boxBtns}>
                    <Button text="Voltar" iconStart={<ArrowLeft />} onClick={handleBack}/>
                    <Button text="PrÃ³ximo" iconEnd={<ArrowRight />} onClick={handleNext} disabled={!state.dentistName} />
                </div>
            </div>
        </div>
    )
}