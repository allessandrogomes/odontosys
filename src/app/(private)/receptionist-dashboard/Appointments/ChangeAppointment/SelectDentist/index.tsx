import styles from "./styles.module.scss"
import Label from "@/components/ui/Label"
import Button from "@/components/ui/Button"
import { ArrowLeft, ArrowRight, OctagonX } from "lucide-react"
import Spinner from "@/components/ui/Spinner"
import useSWR from "swr"
import FeedbackMessage from "@/components/ui/FeedbackMessage"
import fetcher from "@/services/dentist/getDentistByProcedure"
import { useChangeAppointmentContext } from "@/contexts/ChangeAppointmentContext"

export default function SelectDentist() {
    const { state, dispatch } = useChangeAppointmentContext()
    const procedure = state.selectedAppointment?.procedure
    const swrKey = procedure ? `/api/dentist?procedure=${encodeURIComponent(procedure)}` : null
    const { data: dentists, error, isLoading } = useSWR<IDentist[] | null>(swrKey, fetcher, { revalidateOnFocus: false })

    function handleSelect(dentist: IDentist) {
        dispatch({ type: "SET_SELECTED_APPOINTMENT", payload: state.selectedAppointment
            ? {
                ...state.selectedAppointment,
                dentist: dentist,
                dentistId: dentist.id,
                scheduledAt: null,
                endsAt: null
            }
            : null
         })
    }

    function handleBack() {
        dispatch({ type: "SET_STEP", payload: 3 })
    }

    function handleNext() {
        dispatch({ type: "SET_STEP", payload: 5 })
    }

    return (
        <div className={styles.box}>
            <div className={styles.selectView}>
                <Label text="Escolha o Dentista" />
                {isLoading ? <Spinner />
                    : error ? <FeedbackMessage message={error} icon={<OctagonX />} />
                        : (
                            <div className={styles.containerDentists}>
                                {dentists && dentists.map((dentist: IDentist) => (
                                    <div
                                        role="button"
                                        tabIndex={0}
                                        aria-label={`Selecionar dentista ${dentist.name}`}
                                        className={`${styles.dentist} ${state.selectedAppointment && state.selectedAppointment.dentistId === dentist.id ? styles.selected : ""}`}
                                        onClick={() => handleSelect(dentist)}
                                        key={dentist.id}
                                    >
                                        <p>Dr. {dentist.name}</p>
                                    </div>
                                ))}
                            </div>
                        )}
                <div className={styles.boxBtns}>
                    <Button text="Voltar" iconStart={<ArrowLeft />} onClick={handleBack}/>
                    <Button text="PrÃ³ximo" iconEnd={<ArrowRight />} onClick={handleNext} disabled={!state.selectedAppointment?.dentist?.name} />
                </div>
            </div>
        </div>
    )
}