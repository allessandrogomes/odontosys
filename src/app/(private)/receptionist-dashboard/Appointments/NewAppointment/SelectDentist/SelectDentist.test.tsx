import useSWR from "swr"
import SelectDentist from "."
import { fireEvent, render, screen } from "@testing-library/react"
import { useAppointmentContext } from "@/contexts/NewAppointmentContext"


// Mock do contexto
jest.mock("@/contexts/NewAppointmentContext", () => ({
    useAppointmentContext: jest.fn().mockReturnValue({
        state: { procedure: "Limpeza", dentistId: null, dentistName: "" },
        dispatch: jest.fn()
    })
}))

// Mock do SWR
jest.mock("swr", () => ({
    __esModule: true,
    default: jest.fn()
}))

// Mock de subcomponentes complexos
jest.mock("@/components/ui/Button", () => ({
    __esModule: true,
    default: jest.fn(({ text, iconStart, iconEnd, ...props }) => <button {...props}>{iconStart}{text}{iconEnd}</button>)
}))
jest.mock("@/components/ui/Spinner", () => ({
    __esModule: true,
    default: jest.fn(() => <div>Spinner</div>)
}))
jest.mock("@/components/ui/FeedbackMessage", () => ({
    __esModule: true,
    default: jest.fn(({ message }) => <div>{message}</div>)
}))

describe("SelectDentist", () => {
    // Mock global do dispatch
    const dispatchMock = jest.fn()

    beforeEach(() => {
        jest.clearAllMocks(); // Reseta todos os mocks antes de cada novo it

        // Mock do SWR
        (useSWR as jest.Mock).mockReturnValue({ data: [], error: null, isLoading: false })
    })

    it("deve renderizar o componente inicial", () => {
        const { container } = render(<SelectDentist />)
        expect(container).toMatchSnapshot()
    })

    it("deve exibir os dentistas retornados pelo useSWR", () => {
        const dentistMock = [
            { id: "1", name: "João" },
            { id: "2", name: "Maria" }
        ]

            ; (useSWR as jest.Mock).mockReturnValue({ data: dentistMock, error: null, isLoading: false })

        render(<SelectDentist />)

        expect(screen.getByText("Dr. João")).toBeInTheDocument()
        expect(screen.getByText("Dr. Maria")).toBeInTheDocument()
    })

    it("deve mostrar o spinner quando isLoading for true", () => {
        ; (useSWR as jest.Mock).mockReturnValue({ data: [], error: null, isLoading: true })

        const { getByText } = render(<SelectDentist />)

        expect(getByText("Spinner")).toBeInTheDocument()
    })

    it("deve mostrar a mensagem de erro quando error estiver presente", () => {
        ;(useSWR as jest.Mock).mockReturnValue({ data: undefined, error: "Erro inesperado", isLoading: false })

        const { getByText } = render(<SelectDentist />)

        expect(getByText("Erro inesperado")).toBeInTheDocument()
    })

    it("deve armazenar o dentista selecionado no state e habilitar o botão Próximo", () => {
        const dentistMock = [
            { id: 1, name: "João" },
            { id: 2, name: "Maria" }
        ]

        // Contexto inicial sem dentista selecionado
        ;(useAppointmentContext as jest.Mock).mockReturnValue({
            state: { procedure: "Limpeza", dentistId: null, dentistName: null },
            dispatch: dispatchMock
        })

        ;(useSWR as jest.Mock).mockReturnValue({ data: dentistMock, error: null, isLoading: false })

        const { getByLabelText, rerender, getByText } = render(<SelectDentist />)
        
        // Clica no dentista João
        const selectJoao = getByLabelText("Selecionar dentista João")
        selectJoao.click()

        expect(dispatchMock).toHaveBeenCalledWith({
            type: "SET_DENTIST",
            payload: { id: 1, name: "João" }
        })

        // Atualiza o estado simulando que o dentista foi salvo no context
        ;(useAppointmentContext as jest.Mock).mockReturnValue({
            state: { procedure: "Limpeza", dentistId: 1, dentistName: "João" },
            dispatch: dispatchMock
        })

        rerender(<SelectDentist />)

        // Agora o botão Próximo deve estar habilitado
        const nextButton = getByText("Próximo") as HTMLButtonElement
        expect(nextButton).not.toBeDisabled()
    })

    it("deve executar dispatch ao clicar em Voltar e Próximo", () => {
        render(<SelectDentist />)

        const backButton = screen.getByText("Voltar")
        const nextButton = screen.getByText("Próximo")

        // Simula o clique no botão Voltar
        fireEvent.click(backButton)
        expect(dispatchMock).toHaveBeenCalledWith({ type: "SET_STEP", payload: 3 })
        expect(dispatchMock).toHaveBeenCalledWith({
            type: "SET_DENTIST",
            payload: { id: null, name: null }
        })

        // Simula o clique no botão Próximo
        fireEvent.click(nextButton)
        expect(dispatchMock).toHaveBeenCalledWith({ type: "SET_STEP", payload: 5 })
    })
})
