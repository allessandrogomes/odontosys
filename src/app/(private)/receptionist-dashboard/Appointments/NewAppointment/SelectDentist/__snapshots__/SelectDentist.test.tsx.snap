// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`SelectDentist deve renderizar o componente inicial 1`] = `
<div>
  <div
    class="box"
  >
    <div
      class="selectView"
    >
      <label
        class="label"
      >
        Escolha o Dentista
      </label>
      <div
        class="containerDentists"
      />
      <div
        class="boxBtns"
      >
        <button>
          <svg
            aria-hidden="true"
            class="lucide lucide-arrow-left"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m12 19-7-7 7-7"
            />
            <path
              d="M19 12H5"
            />
          </svg>
          Voltar
        </button>
        <button
          disabled=""
        >
          Pr√≥ximo
          <svg
            aria-hidden="true"
            class="lucide lucide-arrow-right"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5 12h14"
            />
            <path
              d="m12 5 7 7-7 7"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;
